lib/src/code_field/code_field.dart

```diff
@@ -393,8 +397,17 @@ class CodeFieldState extends State<CodeField> {
               widget.padding.top -
               _codeScroll!.offset,
           0);
-      painterWidth = painter.width;
-      painterHeight = painter.height;
+      _positionPopup(caretHeight);
     });
   }
+
+  void _positionPopup(double caretHeight) {
+    if (cursorY + Sizes.autocompletePopupMaxHeight > windowSize.height) {
+      cursorY = cursorY -
+          (Sizes.autocompletePopupMaxHeight + caretHeight + Sizes.caretPadding);
```

Comment:
What is `cursorY` defined to store?

Reply:
If we use the preferred offset, then the host widget should only calculate the offset for the default popup position (which is below the cursor) and let the popup widget handle the rest. The host should not decide on the alignment for it. Only then is the SRP observed.

The popup widget will also need the height of the line if it decides to flip.